lib/mayaUsd/render/vp2RenderDelegate/material.cpp

```diff
@@ -113,35 +113,82 @@ inline bool _IsUsdUVTexture(const HdMaterialNode& node)
     return (node.identifier == UsdImagingTokens->UsdUVTexture);
 }
 
-//! Helper utility function to print nodes, connections and primvars in the
+//! Helper function to generate a XML string about nodes, relationships and primvars in the
 //! specified material network.
-void _PrintMaterialNetwork(
-    const std::string&       label,
-    const SdfPath&           id,
-    const HdMaterialNetwork& mat)
+std::string _GenerateXMLString(const HdMaterialNetwork& mat, bool includeParams=true)
```

Comment:
Looking at HdMaterialNode and HdMaterialRelationship, it has members that can return a hash. Did you consider generating hash directly out of objects without having to create a big string and computing a hash for it?

Reply:
Yes, computing the hash from the members is an option but we would have to use the `HdMaterialNetwork` as the key to index the shader cache?

If so, it wouldn't work because the equality comparison between `HdMaterialNetwork` includes the path comparison, making it too strict to support reuse.

Reply:
The material network is a large stream of TfTokens, each one having already a hash computed. So you could probably generate code to combine the hashes in a way that is very similar to that XML representation and use the final hash as key for the shader instance. USD uses boost::hash_combine a lot for these scenarios. This could save quite a few CPU cycles.

Reply:
My concern is that we cannot use hash as the key to index the shader cache because the hash code can collide, meaning two different networks (with different nodes or different relationships) can produce the same hash.

Reply:
I was thinking of using a stripped material network (which removes material paths and params) as the key so that we can use the proposed hash_combine approach, but using it as index also has higher cost than a single-token key. So there is pros & cons for each approach. I would prefer to keep this in mind and when we find the code become a hotspot we explore the above proposal.

Reply:
Maybe use something like std::vector<TfToken> as key then? That would be more compact than generating a large string and the hash_combine strategy would still work.

Reply:
Using std::vector as key is similar to using a stripped material network as key, both would need more equality comparison than single token, while saving compute cost of hash code. I can try this after PR121 based on profiling.